basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepExecution.java

```diff
@@ -22,26 +29,29 @@ public boolean start() throws Exception {
                 .withCallback(new Callback())
                 .withDisplayName(null)  // hide the body block
                 .start();
-        setupTimer();
+        long now = step.getUnit().convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+        end = now + step.getTime();
```

Comment:
Oh, I see what you are trying to do—convert milliseconds into the step’s defined unit—but it is not a good idea. Consider `timeout(time: 1, unit: 'HOURS')` at 4:58 PM. You will lose a lot of precision. And it is very confusing because `end` will effectively be “hours since the epoch”, which is a measure never used anywhere else in Java.

`now` and `end` should rather be defined to be milliseconds since the epoch, and `step.time` converted to milliseconds in `start`. `Timer.schedule` can then hardcode `TimeUnit.MILLISECONDS`. As a bonus, you can then make `step` `transient`, since it would be used only in `start`.